### PR TITLE
fix: 8846: Unresponsive camera opens on iOS

### PR DESCRIPTION
**Pull request by dharmik**
### Fixed Issues

$ [8846](https://github.com/Expensify/App/issues/8846)

### Tests

tests to something as below -

For example:
1. (Make sure you have denied the camera permission)
2. Go to settings > profile
3. Tap on the avatar icon
4. Select Upload photo
5. Select Take photo

- [x] Verify that no errors appear in the JS console

#### Contributor (PR Author) Checklist
- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] iOS / native
    - [x] Android / native
    - [x] iOS / Safari
    - [x] Android / Chrome
    - [x] MacOS / Chrome
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there’s a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained “why” the code was doing something instead of only explaining “what” the code was doing.
    - [x] I verified any copy / text shown in the product was added in all `src/languages/*` files
    - [x] I verified any copy / text that was added to the app is correct English and approved by marketing by tagging the marketing team on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named “index.js”. All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] Any functional components have the `displayName` property
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose and it is
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn’t already exist
    - [x] The style can’t be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.

<details>
<summary><h4>PR Reviewer Checklist</h4></summary>

- [ ] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] iOS / native
    - [ ] Android / native
    - [ ] iOS / Safari
    - [ ] Android / Chrome
    - [ ] MacOS / Chrome
    - [ ] MacOS / Desktop
- [ ] I verified there are no console errors (if there’s a console error not related to the PR, report it or open an issue for it to be fixed)
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained “why” the code was doing something instead of only explaining “what” the code was doing.
    - [ ] I verified any copy / text shown in the product was added in all `src/languages/*` files
    - [ ] I verified any copy / text that was added to the app is correct English and approved by marketing by tagging the marketing team on the original GH to get the correct copy.
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named “index.js”. All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/PR_REVIEW_GUIDELINES.md)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] Any functional components have the `displayName` property
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose and it is broken down into smaller components in order to separate concerns and functions
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn’t already exist
    - [ ] The style can’t be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.

</details>

### QA Steps

tests to something like this -

For example:
1. (Make sure you have denied the camera permission)
2. Go to settings > profile
3. Tap on the avatar icon
4. Select Upload photo
5. Select Take photo

- [x] Verify that no errors appear in the JS console

### Video & Photos

You can also refer bellow screenshots regarding changes

When first time we click on **Take a Photo**
app-camera-permission-alert

When camera permission is denied or blocked
app-permission-setting

For enable Camera permission while tap on Setting
app-camera-permission

Video:
https://drive.google.com/file/d/1RbDU7nGZS_l4SUeyzt7ODePctWD7fnh5/view?usp=sharing

### CODE REVIEW

**Comment by Julesssss** [comment]
Hi @dharmik, could you run `npm run lint` locally and then fix the failing checks please

File: src/components/AttachmentPicker/checkCameraPermission/index.js
Comment (mananjadhav) [inline src/components/AttachmentPicker/checkCameraPermission/index.js:3]:
```suggestion
    * No need to check for Android and other platforms, hence we return true
```

File: src/components/AttachmentPicker/index.native.js
Comment (mananjadhav) [inline src/components/AttachmentPicker/index.native.js:189]: Any specific reason this is resolved as promise and not async await?

Comment (mananjadhav) [reply src/components/AttachmentPicker/index.native.js:189]: I am not sure what we mean by `isValid`. Do we mean `hasPermissions`?

File: src/components/AttachmentPicker/index.native.js
Comment (mananjadhav) [inline src/components/AttachmentPicker/index.native.js:202]: Any reason we changed switch case to if-else here?

File: src/components/AttachmentPicker/index.native.js
Comment (mananjadhav) [inline src/components/AttachmentPicker/index.native.js:187]: I am generally against passing functions as params, especially when they have side-effects. for example in the `else` block no resolve can lead to a wrap when we would want to refactor `this.showPermissionAlert`.

Can get a Promise chain and let `openImagePicker` be an async function or have its own Promise to resolve.

**Review by mananjadhav — CHANGES_REQUESTED** [review]
@dharmik Requested few changes and please attend to the `lint` errors.

Thank you for the PR, I am guessing after the changes are done, we can move towards testing it.

File: src/components/AttachmentPicker/index.native.js
Comment (dharmik) [reply src/components/AttachmentPicker/index.native.js:202]: Resolved.

File: src/components/AttachmentPicker/index.native.js
Comment (dharmik) [reply src/components/AttachmentPicker/index.native.js:187]: I feel it looks okay but let's take some comment from @Julesssss 
CC @mananjadhav

**Comment by dharmik** [comment]
Except this [discussion_r925182318](https://github.com/Expensify/App/pull/9949#discussion_r925182318)
All other comments and lint issues are resolved.
@mananjadhav

**Comment by mananjadhav** [comment]
> Except this [discussion_r925182318](https://github.com/Expensify/App/pull/9949#discussion_r925182318) All other comments and lint issues are resolved. @mananjadhav

@Julesssss Waiting for your quick review before I can move forward with the testing.

File: src/components/AttachmentPicker/checkCameraPermission/index.ios.js
Comment (Julesssss) [inline src/components/AttachmentPicker/checkCameraPermission/index.ios.js]: Can we simplify to the following?

`return result === RESULTS.GRANTED`

File: src/components/AttachmentPicker/checkCameraPermission/index.js
Comment (Julesssss) [inline src/components/AttachmentPicker/checkCameraPermission/index.js]: Minor case change: 

> No need to check for android so **It** return true

> No need to check for android so **it** return true

File: src/components/AttachmentPicker/index.native.js
Comment (Julesssss) [reply src/components/AttachmentPicker/index.native.js:187]: > no resolve can lead to a wrap

@mananjadhav would you mind explaining this part? 

I agree it's not ideal currently. I would be interested too see this change too: 

> Can get a Promise chain and let openImagePicker be an async function or have its own Promise to resolve.

Comment (mananjadhav) [reply src/components/AttachmentPicker/index.native.js:187]: > @mananjadhav would you mind explaining this part?

All I meant here is if we're defining a Promise, lets have either return a Promise chain or call with `resolve`/`reject`

**Comment by mananjadhav** [comment]
Sorry I thought @dharmik must've checked that the PR is not mergeable due to unverified commits.  @dharmik Could you please check this at your end? I can see you've addressed comments but we can't merge this PR if they don't have signed commits. (I learnt it the hard way).

**Comment by Julesssss** [comment]
> Sorry I thought @dharmik must've checked that the PR is not mergeable due to unverified commits. @dharmik Could you please check this at your end? I can see you've addressed comments but we can't merge this PR if they don't have signed commits. (I learnt it the hard way).

Bumping this @dharmik. Unfortunately we can't merge until the commits are signed.

**Comment by dharmik** [comment]
@Julesssss @mananjadhav Sorry but raised another PR because not facing some issue in signed commit.

**Pull request closed without merging**
